### PR TITLE
Centralise logging [RFC]

### PR DESCRIPTION
**Pull request by yschimke**
I keep hitting weirdness in logging when I try to use it, and wanted to tighten up expectations and consistency over the logging approaches.

### CODE REVIEW

File: okhttp/src/main/kotlin/okhttp3/internal/platform/Platform.kt [hunk 1]
```diff
@@ -119,9 +120,12 @@ open class Platform {
     socket.connect(address, connectTimeout)
   }
 
-  open fun log(message: String, level: Int = INFO, t: Throwable? = null) {
-    val logLevel = if (level == WARN) Level.WARNING else Level.INFO
-    logger.log(logLevel, message, t)
+  @Deprecated("Moved to logger field", replaceWith = ReplaceWith("logger.info(message, t)"))
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/internal/platform/Platform.kt:123]: nit: we should never deprecate internal APIs; we can just fix ’em.

File: android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt [hunk 2]
```diff
@@ -297,7 +316,7 @@ class OkHttpTest {
       moshi.adapter(HowsMySslResults::class.java).fromJson(response.body!!.string())!!
     }
 
-    Platform.get().log("results $results", Platform.WARN)
+    Platform.get().getLogger(OkHttpTest::class.java.simpleName).warn("results $results")
```
Comment (swankjesse) [inline android-test/src/androidTest/java/okhttp/android/test/OkHttpTest.kt:319]: I don’t think you want simpleName here?

File: okhttp/src/main/kotlin/okhttp3/internal/platform/Jdk8WithJettyBootPlatform.kt [hunk 3]
```diff
@@ -62,7 +62,9 @@ class Jdk8WithJettyBootPlatform(
     try {
       val provider = Proxy.getInvocationHandler(getMethod.invoke(null, sslSocket)) as AlpnProvider
       if (!provider.unsupported && provider.selected == null) {
-        log("ALPN callback dropped: HTTP/2 is disabled. " + "Is alpn-boot on the boot class path?")
+        logger.debug(
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/internal/platform/Jdk8WithJettyBootPlatform.kt:65]: did the level here change? I think that’s fine but it should be a separate PR

File: okhttp/src/main/kotlin/okhttp3/internal/platform/Platform.kt [hunk 4]
```diff
@@ -119,9 +120,12 @@ open class Platform {
...
+  fun log(message: String, level: Int = INFO, t: Throwable? = null) {
+    when (level) {
+      INFO -> logger.info(message, t)
+      WARN -> logger.warn(message, t)
```
Comment (swankjesse) [inline okhttp/src/main/kotlin/okhttp3/internal/platform/Platform.kt:127]: did debug disappear?

**Comment by swankjesse** [comment]
Maybe I should have started with this question: what kind of weirdness are you running into?

**Comment by swankjesse** [comment]
It’s too bad java.util.Logging misses the mark so badly.

**Comment by yschimke** [comment]
> It’s a nice abstraction but I really hate to introduce our own logging abstraction.

Agreed.  Awkwardly we have some conflicting requirements.

1. We want it to do the right thing by default by two distinct platforms (Android + Java8+)
2. We don't want additional dependencies.
3. Platform.get().log is a useful way of notifying the developer of things in dev mode.  In JVM using JUL achieves this, on Android System.out.println or Log.i do the same.

**Comment by yschimke** [comment]
> I’m grumpy because I just read this which really resonated with me.
> https://sobolevn.me/2020/03/do-not-log

+100 Hopefully this exact topic comes across as a theme I've been pushing for.  A lot of the code in exploratory android work has been trying to build events out of connections, networks etc.

https://github.com/square/okhttp/issues/5194
https://github.com/square/okhttp/issues/5833
https://github.com/square/okhttp/issues/5319

JSSE is the worst example to be honest.  Such a critical API and failures are only debuggable via turning on logging.

I'll raise a clear ticket to cleanup logging and replace with events and observable state.

**Comment by yschimke** [comment]
> Could we use something like OkHttpDebugLogcat so we exclusively use the JUL logging API instead of creating our own?

This is probably the best short term solution, switch to JUL completely, but hook up our own zero dependency logging by default with AndroidPlatform.

**Comment by yschimke** [comment]
> Maybe I should have started with this question: what kind of weirdness are you running into?

It's not fun trying to handle logging in useful way e.g.

https://github.com/yschimke/okurl/blob/master/src/main/kotlin/com/baulsupp/okurl/util/LoggingUtil.kt#L40
